### ISSUE TITLE
IPv6 addresses are compared using case-sensitive string comparisons

### ISSUE THREAD

**Issue by nhinds** [sensu-plugins/sensu-plugins-dns#4]
Given a domain `ipv6.example` resolving to the following address:

```
2001:DB8::1
```

The following checks should pass, but do not:

```
check-dns.rb --domain ipv6.example --type AAAA --result 2001:db8::1
check-dns.rb --domain ipv6.example --type AAAA --result 2001:0db8:0000:0000:0000:0000:0000:0001
check-dns.rb --domain ipv6.example --type AAAA --result 2001:0DB8:0000:0000:0000:0000:0000:0001
check-dns.rb --domain ipv6.example --type AAAA --result 2001:db8:0:0:0:0:0:1
check-dns.rb --domain ipv6.example --type AAAA --result 2001:DB8:0:0:0:0:0:1
check-dns.rb --domain ipv6.example --type AAAA --result 2001:db8::0.0.0.1
check-dns.rb --domain ipv6.example --type AAAA --result 2001:DB8::0.0.0.1
```

The only check that passes is the short form, in uppercase:

```
check-dns.rb --domain ipv6.example --type AAAA --result 2001:DB8:1
```

**Comment by nhinds** [comment]
Debug output looks like:

```
$ /opt/sensu/embedded/bin/check-dns.rb --domain ipv6.example --type AAAA --result 2001:db8:0:0:0:0:0:1 --debug
Entries: ;; Answer received from 10.0.3.1 (86 bytes)
;;
;; Security Level : UNCHECKED
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 35419
;; flags: qr rd ra cd; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

OPT pseudo-record : payloadsize 512, xrcode 0, version 0, flags 32768


;; QUESTION SECTION (1  record)
;; ipv6.example.	IN	AAAA

;; ANSWER SECTION (1  record)
ipv6.example.	252	IN	AAAA	2001:DB8::1
ipv6.example.	252	IN	AAAA	2001:DB8::1
DNS CRITICAL: Resolved ipv6.example AAAA did not include 2001:db8:0:0:0:0:0:1
```

**Comment by jgnagy** [comment]
Once PR #5 is cut as a release, you can solve the case-sensitivity issue using a regular expression search (or, in fact, account for all the scenarios in this specific case) using a sufficiently thorough regular expression like the following:

    check-dns.rb --domain ipv6.example --type AAAA -R '(?i-mx:(2001:(0)?db8:((((000)?0:){5}(000)?1)|:(000)?1|:0.0.0.1)))'

Granted that this is probably not ideal, but the problem is that the check isn't comprehending the DNS query in any meaningful way. It just compares strings. Here's a simple demonstration of your original example using case-insensitive regular expression searching:

    check-dns.rb --domain ipv6.example --type AAAA -R '(?i-mx:(2001:db8::1))'

This will match both `2001:DB8::1` and `2001:db8::1`, but of course won't match `2001:0db8::1` with its extra `0`. I'm not sure if this helps any, but it might solve some of the least troublesome case-sensitivity issues.

As for making the check actually comprehend DNS answers such that it *knows* that `2001:db8::1` is the same as `2001:0DB8:0000:0000:0000:0000:0000:0001`, and so on, that seems like it would take an extensive rewrite using the standard library's `ipaddr` library:

    require 'ipaddr'
    IPAddr.new('2001:db8::1') == IPAddr.new('2001:0DB8:0000:0000:0000:0000:0000:0001')
    => true

**Comment by nhinds** [comment]
This sort of works...

My example is actually a little more complex than I gave in the original description. I've got a fact on each of my nodes (stamped out into my sensu client configuration) representing that node's public IPv6 address, so my actual sensu check has an interpolated value instead of a hardcoded IPv6 address - i.e. my check definition looks like `check-dns.rb --domain ":::name:::" --type AAAA --result ":::public_ipaddress6:::"`.

If I change the parameter to be `--regex-match "(?i):::public_ipaddress6:::"` it seems to match for my use cases - this may be only because dnsruby and my fact happen to output the IPv6 address in a format that only differs by case, but it does make the check pass for all of my hosts.

**Comment by majormoses** [comment]
I think we can certainly ignore the case for string comparison on this.

**Comment by majormoses** [comment]
@nhinds can you check out #18 and validate this works for you?

**Comment by nhinds** [comment]
It fixes half of the problem: instead of checking via case-sensitive string comparison, this is now via a case-insensitive string comparison.

As stated in the description, string comparison is not sufficient for comparing IPv6 addresses. For example, the addresses `2001:db8::1`, `2001:0DB8:0000:0000:0000:0000:0000:0001`, and `2001:DB8::0.0.0.1` are all equivalent, but none of them compare as equal to each other using a string comparison.

**Comment by majormoses** [comment]
gotcha let me see what I can cook up.

**Comment by majormoses** [comment]
Honestly I don't really do much with ipv6 so I would like to check the different scenarios please let me know if I am missing anything or if something is not really valid.

- [ ] ignore case when comparing ipv6 addresses
- [ ] expand ipv6 address comparison to account for shorthand vs longhand
- [ ] support scope id? saw this in windows... `(${SOME_VALID_IPV6_ADDR}%${SCOPE_ID}`
- [ ] loopback

I feel like we only really need to solve the top 2.

**Comment by nhinds** [comment]
I don't believe you can store scope IDs in AAAA records - https://tools.ietf.org/html/rfc3596 says it's just the address.

Loopback ipv6 addresses aren't really different from other addresses - they can just get expanded to their canonical form like any other address.

The best way to support this is to use an IP address library to deal with the comparisons, e.g. [IPAddr](http://ruby-doc.org/stdlib-2.0.0/libdoc/ipaddr/rdoc/IPAddr.html). The lib can handle normalising addresses, ignoring case, expanding addresses, etc.

```
irb(main):001:0> require 'ipaddr'
=> true
irb(main):002:0> IPAddr
=> IPAddr
irb(main):003:0> IPAddr.new('2001:0DB8:0000:0000:0000:0000:0000:0001')
=> #<IPAddr: IPv6:2001:0db8:0000:0000:0000:0000:0000:0001/ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff>
irb(main):004:0> a=_
=> #<IPAddr: IPv6:2001:0db8:0000:0000:0000:0000:0000:0001/ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff>
irb(main):005:0> IPAddr.new('2001:DB8::0.0.0.1')
=> #<IPAddr: IPv6:2001:0db8:0000:0000:0000:0000:0000:0001/ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff>
irb(main):006:0> b=_
=> #<IPAddr: IPv6:2001:0db8:0000:0000:0000:0000:0000:0001/ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff>
irb(main):007:0> c=IPAddr.new('2001:db8::1')
=> #<IPAddr: IPv6:2001:0db8:0000:0000:0000:0000:0000:0001/ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff>
irb(main):008:0> a==b
=> true
irb(main):009:0> a==c
=> true
irb(main):010:0> b==c
=> true
```

**Comment by jgnagy** [comment]
I seem to remember suggesting something like `IPAddr`. :-)

If that's the goal, I can try to take on the work of refactoring the plugin to use `IPAddr` wherever an IP address shows up, unless someone else is interested.

**Comment by majormoses** [comment]
@nhinds yup thats what I was looking at using (I got far enough with the ipaddr lib that I can validate it will work for us but I broke some stuff in the process and need to spend some time refactoring it safely). Thanks for clarifying the scope_id, regarding loopback I meant it's usually only tied to localhost so I was wondering if this was even a valid use case to worry about not that it really adds much if any more complexity. IMHO a lot of the plugins are not in a great state and there has been a lack of maintainer involvement (not blaming anyone just 169 plugin repos and 14 people to moderate and most of us don't work for sensu so it's in our spare time) so I have been trying to prioritize by looking for the lowest hanging fruit first.

@jgnagy certainly not gonna turn away a sane community contribution.
